perf(filetype): implement parent pattern pre-matching

echasnovski opened the pull request:
This PR adds a single commit on top of #29632.

------

Problem: calling `vim.filetype.match()` has performance bottleneck in
  that it has to match a lot of Lua patterns against several versions of
  input file name. This might be the problem if users need to call it
  synchronously a lot of times.

Solution: add "parent pattern pre-matching" which can be used to quickly
  reject several potential pattern matches at (usually rare) cost of
  adding time for one extra Lua pattern match.

  "Parent pattern" is a manually added/tracked grouping of filetype
  patterns which should have two properties:
    - Match at least the same set of strings as its filetype patterns.
      But not too much more.
    - Be fast to match.

  For them to be effective, group should consist from at least three
  filetype patterns.

  Example: for a filetpye pattern ".*/etc/a2ps/.*%.cfg", both "/etc/"
  and "%.cfg" are good parent patterns (prefer the one which can group
  more filetype patterns).

  After this commit, `vim.filetype.match()` on most inputs runs ~3.4
  times faster (while some inputs may see less impact if they match
  many parent patterns).

------

Notes:
- There is more description about the "fast pattern" approach in the new section of 'dev_vimpatch.txt'
- Technically, this can already be used in filetype patterns added via `vim.filetype.add()`, but it is left intentionally undocumented. If it will be decided as a good idea to document, then I'd happily do it (but I'd imagine it is not).
- This is the (almost) the best I can do to optimize performance of `vim.filetype.match()` without breaking compatibility. The only other reasonable route I can think of is to add `skip_slow` flag to `vim.filetype.match()` options which would skip all pattern matching (and possibly content matching). This will not detect filetype for *some* files, but get somewhat faster execution in return (actual speedup is less significant after this PR).

------

Profiling results on some different `filename` inputs (computed in debug builds by measuring 10000 calls of `vim.filetype.match({ filename = 'xxx', buf = buf_id })` with empty buffer and taking median value; executed in '/home/user/repos/neovim' directory; "before fastpat" is recomputed today):

Filename | Before fastpat | After fastpat |
--------------|-----------|--------------|
init.lua | 0.107ms  | 0.030ms |
no-match | 0.128ms | 0.037ms |
/home/user/.config/hello/world/no-match | 0.201ms | 0.059ms |

------

Interesting details:

<details><summary>Motivation on why optimizing pattern matching is the bottleneck</summary>

Here is what `vim.filetype.match()` currently tries (in order):

- Exact match for full name. Fast with table lookup (420 entries).
- Exact match for basename/tail. Fast table lookup (420 entries).
- Pattern match for 355 patterns with positive priority. *Very slow*, as each match is either `string.match()` for _both_ filename as typed and full path (majority of cases) or a single match for basename.
- Extension match for computed extension. Fast table lookup (even with 1092+ entries).
- Similar pattern match for 94 patterns with negative priority. Same method as for positive priority.
- Match based on content.

As matching by extension is (arguably) the most common real world case, reducing execution time of mandatory matching of positive priority patterns is the best way to get overall performance improvement.

</details>

<details><summary>Summary of actually used fast patterns, separately for regular patterns with positive/negative priority and total</summary>

Fast pattern | Positive priority | Negative priority | All patterns |
-------------|-------------------|-------------------|--------------|
'/etc/'      |         79        |         24        |      103     |
''           |         62        |         29        |      91      |
'/log/'      |         77        |         0         |      77      |
'/conf'      |         27        |         1         |      28      |
'%.conf'     |         14        |         10        |      24      |
'/systemd/'  |         17        |         0         |      17      |
'utt'        |         7         |         10        |      17      |
'file'       |         6         |         4         |      10      |
'/%.'        |         8         |         1         |      9       |
'^%.'        |         7         |         2         |      9       |
'/usr/'      |         4         |         5         |      9       |
'git/'       |         6         |         1         |      7       |
'%.cfg'      |         6         |         0         |      6       |
'bash'       |         5         |         0         |      5       |
's6'         |         5         |         0         |      5       |
'/debian/'   |         4         |         0         |      4       |
'/var/'      |         4         |         0         |      4       |
'cmus/'      |         4         |         0         |      4       |
'fvwm'       |         0         |         4         |      4       |
'nginx'      |         4         |         0         |      4       |
'calendar/'  |         0         |         3         |      3       |
'proj%.user$'|         3         |         0         |      3       |
'sst%.meta'  |         3         |         0         |      3       |
'require'    |         3         |         0         |      3       |

</details>

lewis6991 commented:
This seems quite complicated for the speedup. I'm not sure this is justified?

Instead of this convoluted fastpat concept, have you considered restructuring the matching to use a tree instead? E.g. the big block of log patterns can be nested with a parent pattern that covers all of them.

To begin with we could construct this tree manually, but eventually we could generate it.

echasnovski commented:
> Instead of this convoluted fastpat concept, have you considered restructuring the matching to use a tree instead? E.g. the big block of log patterns can be nested with a parent pattern (.*/log/.*) that covers all of them.

That's almost exactly what is done here but instead using "fastpat" option as "parent pattern".

> To begin with we could construct this tree manually, but eventually we could generate it.

This was my initial approach, but there are tricky edge cases. Manual tracking seems to be a more robust approach.

lewis6991 commented:
> but there are tricky edge cases. 

Like what? I would imagine you would need to tokenize the patterns, generate a tree with each node being a single token, and then flatten the tree.

echasnovski commented:
> > but there are tricky edge cases. 
> 
> Like what?

My initial approach was to check if certain set of hand picked patterns can be found as exact substring. This did not 100% work, because they can be followed by modifiers or (in theory) be inside character set (like trying to find 'git/' pattern in a valid '.*[git/]' pattern).

> I would imagine you would need to tokenize the patterns, generate a tree with each node being a single token, and then flatten the tree. 

To me *this* sounds more convoluted. Besides, the core optimization idea is to decrease number of possible pattern matches. This automated approach seems to lead to more overly specific patterns (if I understood what "tokenize the patterns" means) with less performance gain. Having a bit of human touch when picking fast patterns I'd say results in the best performance.

> > That's almost exactly what is done here but instead using "fastpat" option as "parent pattern".
> 
> Almost, but more convoluted. Why not make the parent pattern more explicit, and allow multiple of them.

It is as explicit as it gets - right next to the regular filetype pattern. So instead of creating new nested structures, reuse the current ones (which is a less convoluted solution). It also is "backwards compatible": adding new filetype patterns can be done in exact same way (adding fast patterns is encouraged but is optional).

lewis6991 commented:
> To me this sounds more convoluted. Besides, the core optimization idea is to decrease number of attempts at pattern matching. This automated approach seems to lead to more overly specific patterns (if I understood what "tokenize the patterns" means) with less performance gain. Having a bit of human touch when picking fast patterns I'd say results in the best performance.

I'm not sure that you understand then. Tree decoders are a well understood construct. How do you think your CPU decodes an instruction?

Tokenize means split something like `.*/log/.*` -> `.*` `/` `l` `o` `g` `/` `.*`

Do this for all patterns and create a tree on each of these tokens for all the given patterns.

E.g. for patterns `.*/log/a` and `.*/log/b` you would generate a tree like.

```lua
if c:match('.*') then
  if c:match('.*/') then
    if c:match('.*/l') then
      if c:match('.*/lo') then
        if c:match('.*/log') then
          if c:match('.*/log/') then
            if c:match('.*/log/a') then
              return 'a'
            elseif c:match ('.*/log/b') then
              return 'b'
            end
          end
        end
      end
    end
  end
end
```

You would then flatten this tree to something like:
```lua
if c:match('.*/log/') then
  if c:match('.*/log/a') then
    return 'a'
  elseif c:match ('.*/log/b') then
    return 'b'
  end
end
```

If you have 50 patterns beginning with `.*/log`, you would then only perform 2-50 pattern matches. And this would not need the whole `fastpat` caching thing since you never repeat a match. Therefore, this would be the most performant as it puts less pressure on memory.

You could optimise this further with something like:
```lua
local h, t = c:match('(.*/log/)(.*)')
if h then
  if t:match('^a') then
    return 'a'
  elsif t:match('^b') then
    return 'b'
  end
end
```

But like I said, to begin with we could construct this tree by hand, but generating it would maximise the performance gain as it would minimise the amount of patterns.

> It is as explicit as it gets - right next to the regular filetype pattern. So instead of creating new nested structures, reuse the current ones (which is a less convoluted solution). It also is "backwards compatible": adding new filetype patterns can be done in exact same way (adding fast patterns is encouraged but is optional).

Except you needed to write an entire thesis to describe how it works. I'm sure if you spend enough time studying it in a bubble then it's easy enough to understand, but it is definitely more convoluted than a standard tree decoder.

echasnovski commented:
> Tree decoders are a well understood construct. How do you think your CPU decodes an instruction?
> ...
> Except you needed to write an entire thesis to describe how it works.

@lewis6991, I'd like to publicly say that I would appreciate a slightly less patronizing tone in the future. Ideally, for any author of a good faith PR.

------

In Matrix room discussion there was a consensus that it would benefit to reformat `pattern` table to be nested and indexed first with a fast patterns (instead of each added explicitly on pattern level). I hope to do that shortly.

echasnovski commented:
> In Matrix room discussion there was a consensus that it would benefit to reformat `pattern` table to be nested and indexed first with a fast patterns (instead of each added explicitly on pattern level). I hope to do that shortly.

I've pushed a reformatted `pattern` table. Keys (fast patterns) are ordered by "full directory match" > "directory start" > "directory end" > "plain match". Inside each subtable order is alphabetical by assigned filetype first (as is currently intended on `master`; made a few corrections), then by pattern (alphabetically). Total counts check out, so hopefully did not miss anything :crossed_fingers:

------

@lewis6991, I'd like to share some overnight thoughts about automated "parse pattern and generate tree" approach:
- Removing fastpat caching in favor of nested `if ... then ... else` is complicated by the fact that patterns need to be ordered by priority first. Preordering in built-in table is doable, but users can add their own filetypes with priorities, which complicates things.
- The automated tree folding approach seems to work left to right (or in some type of order), while the manual approach allows grouping based on parts anywhere in the filetype pattern. Like currently is done for '/conf', 'git/', and more.
- Straightforward tree folding might not get the best nesting on these real world cases. Because I imagine it be splitting by '.*/' (which will always match on Linux), or by a letter (like from 'mutt%-.*%-%w+', 'Muttrc.*' patterns). So it would have to have some heuristics at least.

lewis6991 commented on runtime/lua/vim/filetype.lua:
```diff
@@ -1724,522 +1728,614 @@ local filename = {
+--         - Part of a rare enough directory name (like '/conf', 'git/').
+--         - Character sequence reasonably rarely used in real full paths (like 'nginx', 'cmus').
+--
+-- When modifying an existing regular pattern, make sure that it still fits its group.
```
This comment block can be significantly cut down.

echasnovski replied:
Which parts do you think are not necessary?

I tried to make it more compact and this is the result, as any information seems valuable to me. Can not trim down without more guidance.

echasnovski replied:
I've used [your suggestion](https://github.com/neovim/neovim/pull/29660#discussion_r1675665603). Thanks for trimming it down!

lewis6991 commented on runtime/lua/vim/filetype.lua:
```diff
@@ -1724,522 +1728,614 @@ local filename = {
-    -- such as '.*/etc/a2ps/.*%.cfg', '.*enlightenment/.*%.cfg', etc.
-    { priority = -1 },
+  ['/debian/'] = {
+    ['.*/debian/changelog'] = 'debchangelog',
```
The that the outer and inner pattern are different format (anchored vs non-anchored is confusing). Either:

- Make all the outer patterns anchored: `/debian/` -> `.*/debian/.*`
- Make all the inner patterns non-anchored: `.*/debian/changelog` -> `/debian/changelog$`
   - This would also require a change to `vim.filetype.add` to add the `^$`.

echasnovski replied:
I prefer the second one as earlier preliminary testing showed them to be slightly faster. However, @clason expressed concern that it would make porting future filetype changes less straightforward.

If anything, I'd like it to be a separate, more visible in git logs PR.

echasnovski replied:
This was resolved in #29632.

lewis6991 commented on runtime/lua/vim/filetype.lua:
```diff
@@ -2407,39 +2505,54 @@ local function dispatch(ft, path, bufnr, ...)
-      return vim.pesc(vim.env[env])
-    end)
-    if return_early then
+  local pat_cache = pattern_lookup[pat] or parse_pattern(pat)
```
Can we move the pattern parsing to *add* time as opposed to *match* time?

If this has to be deferred until the first autocmd is triggered, then that can be done.

echasnovski replied:
It can be moved to `sort_by_priority()`, as it does some inferring inside of it. However, I am not sure what the benefit of this would be. Could you elaborate, please?

lewis6991 replied:
*match* time is hotter than *add* time. I think it's best we do as much pre-processing at *add* time, and keep matching as simple as possible.

echasnovski replied:
This is resolved in #29632.

lewis6991 commented on runtime/lua/vim/filetype.lua:
```diff
@@ -2449,6 +2562,56 @@ local function match_pattern(name, path, tail, pat)
+--- @param tail string
+--- @param pat string
+--- @return boolean
+local function match_fastpat(name, path, tail, pat)
```
Can this be combined with `match_pattern`?

Unless you can show otherwise, I don't think the additional code is worth the perf gain. And in general I'm not sure we need to introduce a new "fastpat" concept.

echasnovski replied:
It can be combined, but behind an extra flag argument of whether it is regular or fast pattern. The reason for this is that regular filetype matching [depends on whether pattern contains slash or not](https://github.com/neovim/neovim/blob/028dd3c5c4d1828eec64c099d3372ffb90572dc0/runtime/lua/vim/filetype.lua#L2444). If reused direcrly for fast patterns, it would require both types of patterns to share this property, which limits the choice of good fast patterns and needs to be documented (as fragile detail). For example, using 'utt' fast pattern for '.*/%.mutt/muttrc.*' filetype pattern.

Hence there is a separate pattern which tries to match all three possible candidates, independently of slash presence.

echasnovski replied:
> I'm not sure we need to introduce a new "fastpat" concept.

Apart from mentioned more general matching logic, the main reason for "fast pattern" as a concept was that their matching result is cached (hence the name). It is harmful to cache regular filetype matching as it is used at most once.

The general "parse and generate tree" which works like nested `if ... then ... else` *and* also respects priority sorting can indeed remove the need for caching. However it feels like having small extra reward for a single filetype matching stuff compared to the complexity it entails.

lewis6991 replied:
Consider adding an extra flag for now then. I think the whole slash behaviour needs to be revisited as, like priority, is adding significant complexity.

echasnovski replied:
I've added an extra `try_all_candidates` flag with a comment. No more extra function.

lewis6991 commented:
> Removing fastpat caching in favor of nested if ... then ... else is complicated by the fact that patterns need to be ordered by priority first. Preordering in built-in table is doable, but users can add their own filetypes with priorities, which complicates things.

Depends how it is done. User added filetypes just need to be inserted at the appropriate place in the internal table. As opposed to fully resorting like we do now.

> The automated tree folding approach seems to work left to right (or in some type of order), while the manual approach allows grouping based on parts anywhere in the filetype pattern. Like currently is done for '/conf', 'git/', and more.

There are many ways it can work. Left to right would just be the most naive. Nothing suggests that the manual groups added in this PR are more optimal than an automated left-to-right approach. In fact, any non-match with these groups will incur ~20 match operations, which is more than I would expect from a fully optimised tree decoder.

- Straightforward tree folding might not get the best nesting on these real world cases. Because I imagine it be splitting by './' (which will always match on Linux), or by a letter (like from 'mutt%-.%-%w+', 'Muttrc.*' patterns). So it would have to have some heuristics at least.

The majority of these patterns are very niche, so in the real world this table will rarely match. Heuristics will be needed if we want to optimize `Muttrc`, but I do think we need to at all. Niche cases are allowed to be "slow".

> Because I imagine it be splitting by './' (which will always match on Linux)

No, not at all. The splitting will be at whatever point it needs to split. `%.` and `/` aren't treated specially.

lewis6991 commented on runtime/lua/vim/filetype.lua:
```diff
@@ -1724,522 +1728,614 @@ local filename = {
+local detect_xkb = starsetf('xkb', { fastpat = '/usr/' })
 
---- @type vim.filetype.mapping
+-- Table of filetype pattern matching rules grouped by their "fast pattern"
+--
+-- Every filetype pattern match is prefaced with a (cached) matching of its parent "fast pattern".
+-- If there is no match, skip all matching inside group. For that to work, fast pattern should:
+-- - Match at least the same set of strings as filetype patterns inside it. But not too much more.
+-- - Be fast to match (will be matched directly with `string.match(target, fastpat)`).
+--
+-- The core idea behind speedup is that matching results of fast patterns are cached and reused
+-- inside single same `M.match()` call. The effect is as big as the quality of compromise between:
+-- - Fast pattern groups many filetype patterns: leads to better cache reuse in case of no match.
+-- - Fast pattern is specific: leads to less "fast pattern matches but regular one does not" which
+--   has the penalty of a (not useful) single extra fast pattern match.
+--
+-- Example:
+-- - Regular pattern: '.*/etc/a2ps/.*%.cfg'.
+-- - Good fast patterns: '/etc/'; '%.cfg$' (depends on how many patterns they can group)
+-- - Bad fast patterns: '%.' (fast but not specific), '/a2ps/.*%.' (slow but specific)
+--
+-- When adding a new filetype with pattern matching rule, consider the following:
+-- - If there is already a group with appropriate fast pattern, use it.
+-- - If there can be a fast and specific enough pattern to group at least 3 filetype patterns,
+--   add it as a separate grouped entry. Good new fast pattern should be:
+--     - Fast. Good rule of thumb is that it should be a short specific string (i.e. no quantifiers
+--       or character sets).
+--     - Specific. Good rules of thumb (from best to worst):
+--         - Full directory name (like '/etc/', '/log/').
+--         - Part of a rare enough directory name (like '/conf', 'git/').
+--         - Character sequence reasonably rarely used in real full paths (like 'nginx', 'cmus').
+--
+-- When modifying an existing regular pattern, make sure that it still fits its group.
```
```suggestion
-- Table of filetype pattern matching rules grouped by their parent pattern.
--
-- Every filetype pattern match is prefaced with a matching of its parent pattern.
-- If there is no match, skip all matching inside group.
--
-- When modifying an existing regular pattern, make sure that it still fits its group.
--
-- Note that unlike leaf patterns, parent patterns do not have special matching behaviour if they contain a `/`.
```

clason replied:
My rule of thumb: Documentation should explain usage; rationale belongs into the commit message. Rules of thumb in this case are better placed in CONTRIBUTING (for porting filetype.vim patches) or `:h vim.filetype.add` (for user patterns).

echasnovski replied:
I've used the patch (with slightly rearranged sentences for better grouping), thanks! 

The (compressed) rest of it was added as the new section to 'dev_vimpatch.txt', as this is where 'CONTRIBUTING.md' eventually links.

echasnovski committed:
perf(filetype): implement parent pattern pre-matching

Problem: calling `vim.filetype.match()` has performance bottleneck in
  that it has to match a lot of Lua patterns against several versions of
  input file name. This might be the problem if users need to call it
  synchronously a lot of times.

Solution: add "parent pattern pre-matching" which can be used to quickly
  reject several potential pattern matches at (usually rare) cost of
  adding time for one extra Lua pattern match.

  "Parent pattern" is a manually added/tracked grouping of filetype
  patterns which should have two properties:
    - Match at least the same set of strings as its filetype patterns.
      But not too much more.
    - Be fast to match.

  For them to be effective, group should consist from at least three
  filetype patterns.

  Example: for a filetpye pattern ".*/etc/a2ps/.*%.cfg", both "/etc/"
  and "%.cfg" are good parent patterns (prefer the one which can group
  more filetype patterns).

  After this commit, `vim.filetype.match()` on most inputs runs ~3.4
  times faster (while some inputs may see less impact if they match
  many parent patterns).

echasnovski commented:
@gpanders, @clason, @lewis6991: I've force pushed some suggested changes:
- There is no separate `match_fastpat()`. The existing one is reused with a flag.
- Big "thesis" comment is cut down.

Also rebased on the latest #29632, as it contains resolution to some suggestion from this PR.

echasnovski commented:
Only now noticed broken Windows test about symlinks. Turns out it was due to typo (`name:match(path)` instead of `name:match(path)`). Sigh... Should be all good now.

gpanders commented on runtime/doc/dev_vimpatch.txt:
```diff
@@ -302,4 +302,40 @@ used in new documentation:
+==============================================================================
+FILETYPE DETECTION                 *dev-vimpatch-filetype*
+
+Nvim's [filetype detection](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/filetype.lua)
```
Nit: This is Markdown syntax, but this is a Vim help buffer. Probably fine, but I think we typically would put the URL on a separate line with a "See " prefix or something.

And since this is linking to a a file in Nvim's runtime we could just use a file path which will let people easily jump to the file inside of Nvim with `gf`.

```suggestion
See $VIMRUNTIME/runtime/lua/vim/filetype.lua
```

echasnovski replied:
Addressed f47dae74398c2db981c8266494c2035953227489 with `$VIMRUNTIME/lua/vim/filetype.lua` path.

gpanders commented on runtime/lua/vim/filetype.lua:
```diff
@@ -2230,29 +2293,36 @@ local function compare_by_priority(a, b)
 end
 
---- @param t vim.filetype.mapping
+local parse_pattern = function(pat)
```
```suggestion
--- @param pat string
local function parse_pattern(pat)
```

echasnovski replied:
Addressed 168d9e07fd855a2ef33bbe922b2295d9a9e7c3dd (with even and a bit more :) )

gpanders commented on runtime/lua/vim/filetype.lua:
```diff
@@ -2444,24 +2516,35 @@ local function expand_envvar_pattern(pat)
   return some_env_missing, expanded
 end
 
+--- Cache for whether current candidates match their parent patterns.
+--- Need to be manually reset in every new `M.match()` call.
+--- @type table<string,boolean>
+local parent_matches = {}
```
Is there a reason to make this table scoped to the entire module and reset manually in `match()`, rather than just allocating the table in `match()` and passing it as an argument to `match_pattern_sorted`? If we can, we should try to keep the scope as narrow as possible.

echasnovski replied:
No reason apart from not having to make a six argument function. Will make it an argument.

echasnovski replied:
Addressed in 168d9e07fd855a2ef33bbe922b2295d9a9e7c3dd.

echasnovski commented:
Rebased on latest `master` and added new `['%.opam%.locked$'] = 'opam'` pattern.

echasnovski commented:
Thanks for merging! This also unblocks some other potential improvements.

justinmk commented on runtime/doc/dev_vimpatch.txt:
```diff
@@ -302,4 +302,40 @@ used in new documentation:
+      But not too much more.
+    - Be fast to match.
+
+  When adding a new filetype with pattern matching, consider the following:
+  - If there is already a group with appropriate parent pattern, use it.
+  - If there can be a fast and specific enough pattern to group at least
+    3 filetype patterns, add it as a separate grouped entry.
```
This seems like a house of cards. What happens if people get this wrong? Is adding a new filetype now 5x harder than it used to be? Is this worth the gain?

echasnovski replied:
People can always opt for adding new built-in pattern (which is a quite rare occasion) to ['' parent pattern](https://github.com/neovim/neovim/blob/6d997f8068a89703823f1572c56a6331c9e024aa/runtime/lua/vim/filetype.lua#L2298). This will still work but without any performance gains.

Yes, this is very much worth the 3x-4x gain on common cases and is easily visible when used in real world cases inside plugins. Here is [one example](https://github.com/ibhagwan/fzf-lua/issues/1358#issuecomment-2254174945).
